### PR TITLE
Resource refactorings

### PR DESCRIPTION
**Pull request by stefanobaghino-da**
- Add `unreleasable` method to create a `Resource` from a plain `Future` without having to create an owner around it (not sure about the name, accepting suggestions)
- Extract methods that don't really create a `Resource` but are just wrapped into one for _monadic_ reasons (and wrap them just once)
- Make `Resource` and `ResourceOwner` covariant and let the type system handle variance

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/apiserver/StandaloneApiServer.scala [hunk 1]
```diff
@@ -134,7 +133,7 @@ final class StandaloneApiServer(
         List(AuthorizationInterceptor(authService, ec)),
         metrics
       )(actorSystem, materializer, logCtx).acquire()
-      _ <- ResourceOwner.forFuture(() => writePortFile(apiServer.port)).acquire()
+      _ <- Resource.unreleasable(writePortFile(apiServer.port))
```
Comment (SamirTalwar-DA) [inline ledger/sandbox/src/main/scala/com/digitalasset/platform/apiserver/StandaloneApiServer.scala]: Perhaps `Resource.fromFuture` reads better?

File: libs-scala/resources/src/main/scala/com/digitalasset/resources/Resource.scala [hunk 2]
```diff
@@ -40,37 +40,31 @@ trait Resource[A] {
...
   import scala.language.higherKinds
 
-  def apply[T](future: Future[T], releaseResource: T => Future[Unit])(
-      implicit executionContext: ExecutionContext
-  ): Resource[T] =
-    apply(future, releaseResource, () => Future.successful(()))
-
-  private def apply[T](
-      future: Future[T],
+  private def nest[T](future: Future[T])(
```
Comment (SamirTalwar-DA) [inline libs-scala/resources/src/main/scala/com/digitalasset/resources/Resource.scala]: Good name.

File: libs-scala/resources/src/main/scala/com/digitalasset/resources/ResourceOwner.scala [hunk 3]
```diff
@@ -13,7 +13,7 @@ import scala.language.higherKinds
 import scala.util.{Failure, Success, Try}
 
 @FunctionalInterface
-trait ResourceOwner[A] {
+trait ResourceOwner[+A] {
```
Comment (SamirTalwar-DA) [inline libs-scala/resources/src/main/scala/com/digitalasset/resources/ResourceOwner.scala]: I'm really glad you figured out how to make this work.

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/apiserver/StandaloneApiServer.scala (hunk 1 above)
Comment (stefanobaghino-da) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/apiserver/StandaloneApiServer.scala]: Addressed by https://github.com/digital-asset/daml/pull/4382/commits/107e162c8d59dbb6d113a47cd9e9b597f8b760d2

**Pull request merged**